sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/FilterEstimationSuite.scala

```diff
@@ -0,0 +1,309 @@
+      Some(7L)
+    )
+  }
+
+  test("cdate = '2017-01-02' ") {
+    // the predicate is: WHERE cdate = "2017-01-02"
+    val d20170102 = Date.valueOf("2017-01-02")
+    validateEstimatedStats(
+      arDate,
+      Filter(EqualTo(arDate, Literal("2017-01-02")), ChildStatsTestPlan(Seq(arDate))),
+      ColumnStat(distinctCount = 1, min = Some(d20170102), max = Some(d20170102),
+        nullCount = 0, avgLen = 4, maxLen = 4),
+      Some(1L)
+    )
+  }
+
+  test("cdate = cast('2017-01-02' AS DATE)") {
+    // the predicate is: WHERE cdate = cast("2017-01-02" AS DATE)
+    val d20170102 = Date.valueOf("2017-01-02")
+    val d20170102_SQLDate = DateTimeUtils.fromJavaDate(d20170102)
+    validateEstimatedStats(
+      arDate,
+      Filter(EqualTo(arDate, Literal(d20170102_SQLDate, IntegerType)),
+        ChildStatsTestPlan(Seq(arDate))),
+      ColumnStat(distinctCount = 1, min = Some(d20170102), max = Some(d20170102),
+        nullCount = 0, avgLen = 4, maxLen = 4),
+      Some(1L)
+    )
+  }
+
+  test("cdate < '2017-01-03' ") {
+    // the predicate is: WHERE cdate < "2017-01-03"
+    val d20170103 = Date.valueOf("2017-01-03")
+    validateEstimatedStats(
+      arDate,
+      Filter(LessThan(arDate, Literal("2017-01-03")), ChildStatsTestPlan(Seq(arDate))),
+      ColumnStat(distinctCount = 2, min = Some(dMin), max = Some(d20170103),
+        nullCount = 0, avgLen = 4, maxLen = 4),
+      Some(3L)
+    )
+  }
+
+  test("ctimestamp = '2017-01-01 02:00:00' ") {
+    // the predicate is: WHERE ctimestamp = "2017-01-01 02:00:00"
+    val ts2017010102 = Timestamp.valueOf("2017-01-01 02:00:00")
+    validateEstimatedStats(
+      arTimestamp,
+      Filter(EqualTo(arTimestamp, Literal("2017-01-01 02:00:00")),
+        ChildStatsTestPlan(Seq(arTimestamp))),
+      ColumnStat(distinctCount = 1, min = Some(ts2017010102), max = Some(ts2017010102),
+        nullCount = 0, avgLen = 8, maxLen = 8),
+      Some(1L)
+    )
+  }
+
+  test("ctimestamp < '2017-01-01 03:00:00' ") {
+    // the predicate is: WHERE ctimestamp < "2017-01-01 03:00:00"
+    val ts2017010103 = Timestamp.valueOf("2017-01-01 03:00:00")
+    validateEstimatedStats(
+      arTimestamp,
+      Filter(LessThan(arTimestamp, Literal("2017-01-01 03:00:00")),
+        ChildStatsTestPlan(Seq(arTimestamp))),
+      ColumnStat(distinctCount = 2, min = Some(tsMin), max = Some(ts2017010103),
+        nullCount = 0, avgLen = 8, maxLen = 8),
+      Some(3L)
+    )
+  }
+
+  def ChildStatsTestPlan(outList: Seq[Attribute]): StatsTestPlan = {
+    StatsTestPlan(
+      outputList = outList,
+      rowCount = 10L,
+      attributeStats = AttributeMap(Seq(
+        arInt -> childColStatInt,
+        arDate -> childColStatDate,
+        arTimestamp -> childColStatTimestamp
+      ))
+    )
+  }
+
+  private def validateEstimatedStats(
+      ar: AttributeReference,
+      filterNode: Filter,
+      expectedColStats: ColumnStat,
+      rowCount: Option[Long] = None)
```

Comment:
We may want to expand to include some error cases.  Then we will return None for expectedRowCount.